### DISCUSSION TITLE
Crossing a pre-signal should reserve the exit block

### DISCUSSION THREAD

**Discussion by linaclogic** [voxeltycoon/issues#826] [General]
Suggestion: When passing a pre-signal trains should reserve one additional block past the exit signal.

Why?
When a train crosses a pre-signal it only reserves the block between pre-signal and the next signal. If that next signal turns red while the train is travelling toward it, it will get stuck in the block between pre-signal and main signal. But that is what a pre-signal is supposed to prevent from happening.

Example:
A two-way station that is currently occupied, with trains waiting to go in on either end:

![trains1](https://user-images.githubusercontent.com/75800362/120076983-6bc45600-c0a8-11eb-9180-7ec0d9e45ccb.jpg)

So far the pre-signals are working and the trains are waiting in the correct position. Now one of the trains leaves the station. At that moment, both waiting trains will start to move toward the same track:

![trains2](https://user-images.githubusercontent.com/75800362/120076988-70890a00-c0a8-11eb-968b-33e3f5e009de.jpg)

Result: the trains are stuck despite correct signal placement.

**Comment by somebadcode** [comment]
The signalling system leaves a lot to wish for and a few tweaks here and there won't help. OpenTTD suffers from bad signaling because it has been designed from wrong the start.

I have experience operating trains in two different countries that have completely different signalling systems.

Some tips on how to solve signals is to give signals the same kind of rules that are found in real life. For anything realistic to be able to take place, the trains must get reserved a path up until a point where a deadlock is guaranteed to not occur. The train does make decisions, a dispatcher (player) or computer decides. Trains must be dumb and make nothing other than follow the itinerary and follows movement authority. The AI must not make trains autonomous like cars. The AI controls the track circuits and avoids deadlocks by reserving paths for trains. Reserving paths and solving the path finding centrally to take all trains into account is the only proper way to reach any form of realism and efficiency.

There are some things to keep in mind to make the path finding and reservation work:

1. The back of a signal has absolutely no meaning to a train but the signal itself still marks the border of a track circuit. A track circuit does not have to have a signal. If a signal is missing, a different signal prior to the circuit's border gives the movement authority (this would help in the scenario you mention). So track circuits are joined when a movement is reserved.
2. A signal only signals movement authority to the engineer and shows the state of the movement reservation.
3. A movement authority is valid up until the next signal, which can include multiple track circuits.
4. Track circuits are reserved and locked until the train has passed. This can be undone manually by the dispatcher (player).
5. Blocks and track circuits are grouped together with all adjacent blocks, forming a continuous path between stations or junctions.
6. A path has a direction and signals MUST NOT show conflicting movement authorities.
7. A path's block signals automatically shows the green aspect for the direction the path has been set to unless a track circuit is occupied. The opposite direction always shows the red aspect.
8. A path starts with a block signal.
9. The first block signal can be called something along the lines of "an exit block signal".
10. An exit block signal shows the red aspect until a movement authority is explicitly given.
11. A path ends with an entry signal which is not part of the path.
12. An entry signal shows the red aspect until a movement authority is explicitly given.
13. An entry signal gives movement authority up to the next entry signal, exit signal, exit block signal or buffer stop. This is the most difficult part in a signalling system.
14. Multiple concurrent movement authorities can be given for a junction as long as they don't intersect each others' track circuits.
15. Movement authority for entering a station or function is only given if it can guarantee that the train can reach the platform and a reservation for turning around or proceeding must be made beforehand.
16. Paths between stations and junctions have a default direction which a player chooses. For the direction to be changed, the player must change it or the AI must be able to conclude a guarantee that the train can pass through and leave the path empty so the path's direction can be restored, before the direction can be swapped.
17. There are implicit track circuits at the edge of platforms and at all ends of a switch.

For inspiration, developers can look at ERTMS with ETCS level 2 and just think of signals as eye candy and where the RBC makes the decisions and this is where some kind of AI fits in.

Examples of aspects:

- Red: Danger
- Green: Proceed (next signal can be red), possible for all signals.
- Two green: Proceed through next section (next signal shows at least one green), possible for all signals.
- Yellow with green: Movement authority is only for entering into the station/junction.
- Yellow with red: Danger, but entry or exit signal (not block or exit block signal). Same as normal danger.

Fun effects:

1. Flashing red means that the player has manually overridden the signal and the train will continue at very low speed to the next signal in the reserved path. This requires that a reservation has been made and queued. Low speed to allow the player to halt the train.
2. Track circuits that have no signals can have a dwarf signal that only shows basic aspects such as red or green.
